File: data_structure/Deque/java/Deque.java
package data_structure.deque;

import data_structure.linked_list.doubly_linked_list.DoublyLinkedList;
import data_structure.linked_list.doubly_linked_list.Node;

public class Deque <T> {
    DoublyLinkedList<T> list;

    public Deque() {
        list = new DoublyLinkedList<>();
    }

    /**
     * 모든 요소 출력
     */
    public void printAll() {
        list.printAll();
    }

    /**
     * 헤드에 값 삽입
     */
    public void addFirst(T data) {
        list.insertAt(0, data);
    }

    /**
     * 헤드 값 삭제
     */
    public Node<T> removeFirst() {
        return list.deleteAt(0);
    }

    /**
     * 테일 값 삽입
     */
    public void addLast(T data) {
        list.insertAt(list.getCount(), data);
    }

    /**
     * 테일 값 삭제
     */
    public Node<T> removeLast() {
        return list.deleteLast();
    }

    /**
     * 리스트가 비었는지 체크
     */
    public boolean isEmpty() {
        return list.getCount() == 0;
    }
}

File: theory/data_structure/Deque/python/deque.py
import sys, os
sys.path.append(os.path.dirname(os.path.abspath(os.path.dirname(__file__))))

from linked_list import doubly_linked_list
DoublyLinkedList = doubly_linked_list.DoublyLinkedList


class Deque:
    def __init__(self):
        self.list = DoublyLinkedList()

    def print_all(self):
        self.list.print_all()

    def add_first(self, data):
        self.list.insert_at(0, data)

    def remove_first(self):
        return self.list.delete_at(0)

    def add_last(self, data):
        self.list.insert_at(self.list.count, data)

    def remove_last(self):
        return self.list.delete_last()

    def is_empty(self):
        return self.list.count == 0
